File: TypeScript/Array and Strings/Easy/findSecondMax.ts
function findSecondMax(nums: number[]): number {
    const len = nums.length;
    nums.sort();
    const max = nums[len - 1];
    let secondMax = nums[0];

    for (let i = 0; i < len; i++) {
        if (nums[i] > secondMax && nums[i] < max)
            secondMax = nums[i];
    }
    return secondMax;
}

console.log(findSecondMax([]));
console.log(findSecondMax([1]));
console.log(findSecondMax([1, 2]));
console.log(findSecondMax([1, 2, 1, 2, 3, 1, 1]));
console.log(findSecondMax([1, 2, 4, 3, 2, 1, 2, 6, 6, 6, 6]));

File: C/Arrays/find-second-max.c
#include <stdio.h>
#include <limits.h>

int findSecondMax(int *arr, int size) {
    if (size < 2)
        return -1;

    int max = INT_MIN;
    int secondMax = INT_MIN;

    for (int i = 0; i < size; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }

    for (int i = 0; i < size; i++) {
        if ((arr[i] > secondMax) && (arr[i] < max)) {
            secondMax = arr[i];
        }
    }

    // no second maximum
    if (secondMax == INT_MIN) {
        return -1;
    }

    return secondMax;
}

int main() {
    int arr[] = {-1, -1 , -1};
    int size = sizeof(arr) / sizeof(arr[0]);

    printf("%d\n", findSecondMax(arr, size));
    return 0;
}